core/new-gui/src/app/workspace/service/workflow-graph/model/workflow-graph.ts

```diff
@@ -0,0 +1,349 @@
+// type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
+
+/**
+ * WorkflowGraphReadonly is a type that only contains the readonly methods of WorkflowGraph.
+ *
+ * Methods that could alter the graph: add/delete operator or link, set operator property
+ *  are omitted from this type.
+ */
+export type WorkflowGraphReadonly = Omit<WorkflowGraph, restrictedMethods>;
+
+/**
+ * WorkflowGraph represents the Texera's logical WorkflowGraph,
+ *  it's a graph consisted of operators <OperatorPredicate> and links <OpreatorLink>,
+ *  each operator and link has its own unique ID.
+ *
+ */
+export class WorkflowGraph {
+
+  private readonly operatorIDMap = new Map<string, OperatorPredicate>();
+  private readonly operatorLinkMap = new Map<string, OperatorLink>();
+
+  private readonly operatorAddSubject = new Subject<OperatorPredicate>();
+  private readonly operatorDeleteSubject = new Subject<{ deletedOperator: OperatorPredicate }>();
+  private readonly linkAddSubject = new Subject<OperatorLink>();
+  private readonly linkDeleteSubject = new Subject<{ deletedLink: OperatorLink }>();
+  private readonly operatorPropertyChangeSubject = new Subject<{ oldProperty: object, operator: OperatorPredicate }>();
+
+  constructor(
+    operatorPredicates: OperatorPredicate[] = [],
+    operatorLinks: OperatorLink[] = []
+  ) {
+    operatorPredicates.forEach(op => this.operatorIDMap.set(op.operatorID, op));
+    operatorLinks.forEach(link => this.operatorLinkMap.set(link.linkID, link));
+  }
+
+  /**
+   * Adds a new operator to the graph.
+   * Throws an error the operator has a duplicate operatorID with an existing operator.
+   * @param operator OperatorPredicate
+   */
+  public addOperator(operator: OperatorPredicate): void {
+    this.assertOperatorNotExists(operator.operatorID);
+    this.operatorIDMap.set(operator.operatorID, operator);
+    this.operatorAddSubject.next(operator);
+  }
+
+  /**
+   * Deletes the operator from the graph by its ID.
+   * Throws an Error if the operator doesn't exist.
+   * @param operatorID operator ID
+   */
+  public deleteOperator(operatorID: string): void {
+    const operator = this.getOperator(operatorID);
+    if (!operator) {
+      throw new Error(`operator with ID ${operatorID} doesn't exist`);
+    }
+    this.operatorIDMap.delete(operatorID);
+    this.operatorDeleteSubject.next({ deletedOperator: operator });
+  }
+
+  /**
+   * Returns whether the operator exists in the graph.
+   * @param operatorID operator ID
+   */
+  public hasOperator(operatorID: string): boolean {
+    return this.operatorIDMap.has(operatorID);
+  }
+
+  /**
+   * Gets the operator with the operatorID.
+   * Throws an Error if the operator doesn't exist.
+   * @param operatorID operator ID
+   */
+  public getOperator(operatorID: string): OperatorPredicate | undefined {
+    return this.operatorIDMap.get(operatorID);
+  }
+
+  /**
+   * Returns an array of all operators in the graph
+   */
```

Comment:
getAllOperators? Similarly, name the function below as getAllLinks?

Reply:
fixed